### ISSUE TITLE
External CSS files loading multiple times

### ISSUE THREAD

**Issue by frxnz** [ausi/cq-prolyfill#27]
I'm using Webpack and the Sass mixin, and I'm having files get reloaded 6+ times. Am I wrong that passing `postcss` shouldn't reload external files at all?

I'm loading the polyfill like so:

``` js
var cq = require('cq-prolyfill')({ postcss: true });
```

And then calling

``` js
cq.reevaluate(false);
```

Is there an extra step I need to do, or something I'm missing?

**Comment by ausi** [comment]
You are right, with `postcss: true` CSS files shouldn’t get reloaded at all. The only exception is if the CSS files are loaded from a different origin, take a look at the [cross origin stylesheets documentation](https://github.com/ausi/cq-prolyfill/blob/master/docs/cors.md) for the correct configuration for this case.

Calling `reevaluate()` is not necessary in most cases.

Can you provide a link to the affected webpage or a small test case that reproduces this issue?

**Comment by frxnz** [comment]
The project I'm working  on is an embedded application, so I have a JavaScript file loading an additional stylesheet. This is why I'm calling `reevaluate`. The test site I'm embedding on does have some other external stylesheets ( Google fonts and Font Awesome), which happen to be the two that are getting reloaded over and over. 

I don't have a public link to share, but I can try to get a reduced test case done tomorrow.

**Comment by ausi** [comment]
The problem with external stylesheets is, that the prolyfill cannot access the style rules of it and then tries to load them via ajax. But if the same stylesheet gets loaded multiple times that is definitely an issue.

**Comment by frxnz** [comment]
Ok I've got a very basic test case that reproduces the issue I'm having here: http://codepen.io/frxnz/pen/QEQdPW

You can see `font-awesome.min.css` and `bootstrap.min.css` get loaded 3 or 4 times
![image](https://cloud.githubusercontent.com/assets/1072711/16917652/7cf0d6c0-4cd0-11e6-8b88-62457d64536d.png)

I didn't see it in the docs, but is there any way to specify files that should be ignored altogether?

**Comment by ausi** [comment]
Thanks for the basic test case! I will try to fix this the next time I’m working on it.

> is there any way to specify files that should be ignored altogether?

No, currently all stylesheets get processed, but such a configuration option could make sense IMO.

**Comment by developit** [comment]
I had come here to open an issue asking if you'd be up for adding a configuration option for a whitelist/blacklist of CSS URLs to parse/skip.  We have probably 50 stylesheets on a page I can't get rid of (yet) and cq-prolyfill is requesting them all multiple times despite the fact that we only have PostCSS-processed container queries in one of the stylesheets.  

Do you think it'd be worth opening a new issue for a whitelist option?  Also, on a related note - it'd be really neat to have a module (and webpack loader?) that would do the rule parsing and just return the rule+query mappings so that the runtime parsing could be skipped entirely.

**Comment by ausi** [comment]
> cq-prolyfill is requesting them all multiple times despite the fact that we only have PostCSS-processed container queries in one of the stylesheets.

Did you use the configuration `postcss: true` and [enabled CORS](https://github.com/ausi/cq-prolyfill/blob/master/docs/cors.md) for the stylesheets if they are loaded from a different domain? If so, cq-prolyfill shouldn’t request the stylesheets at all.

> Do you think it'd be worth opening a new issue for a whitelist option?

Yes, please open an issue for that.

> it'd be really neat to have a module (and webpack loader?) that would do the rule parsing and just return the rule+query mappings so that the runtime parsing could be skipped entirely.

Please open a separate issue for that too.

**Comment by developit** [comment]
They're on the same domain and `postcss` is set to `true`, yup.  I'll open those issues quickly so we can discuss.

**Comment by ausi** [comment]
> They're on the same domain and `postcss` is set to `true`, yup.

That looks like a bug, but I was not able to reproduce it. Can you provide a link to the affected page or to a testcase?

**Comment by developit** [comment]
Can't provide a link, sadly, as the product is not public yet.  I can look into a test case once I manage to scrap some time together.  From my debugging, I recall that [this flag](https://github.com/ausi/cq-prolyfill/blob/master/cq-prolyfill.js#L101) was for some reason never being set such that the parsing would be bypassed.  I'm calling `reevaluate(false, null, el)` - perhaps that's triggering some odd behavior?

**Comment by ausi** [comment]
The third parameter of `reevaluate()` should be an array of DOM nodes ([API documentation](https://github.com/ausi/cq-prolyfill/blob/master/docs/api.md#reevaluatebool-clearcache-fn-callback-array-contexts)). The `processed` flag should be OK.

The actual bypassing of the stylesheet download is in [cq-prolyfill.js:276-287](https://github.com/ausi/cq-prolyfill/blob/ec5006501c575b1012edcfa6592270c772e5e2b2/cq-prolyfill.js#L276-L287). So as long as `config.postcss` is true and sheet.cssRules is accessible, the stylesheet should not get downloaded.

**Comment by developit** [comment]
Ah yes, I was in fact passing an Array. If I'm seeing `.cssRules` be `null`, is that indicative of a CORS issue?

**Comment by ausi** [comment]
Yes, `cssRules` could be `null` due to a CORS issue, but if all stylesheets are on the same domain as the webpage CORS issues should not appear.

**Comment by developit** [comment]
Might be on a static domain. TIL!

**Comment by ausi** [comment]
Fixed with #36

**Comment by developit** [comment]
Awesome, thanks for your hard work @ausi :)
